File: serial/KGM-JS/Edge.cpp
/* 
 * File:   Edge.cpp
 * 
 */

#include <iostream>

#include "Edge.h"

Edge::Edge(int vertex1, int vertex2) {
	this->vertex1 = vertex1;
	this->vertex2 = vertex2;
}

Edge::~Edge() {
}

bool Edge::isBacktrackMarker() {
	return (vertex1 == -1 && vertex2 == -1);
}

std::ostream & operator <<(std::ostream & os, const Edge & e) {
	int edgeFrom, edgeTo;
	if (e.vertex1 <= e.vertex2) {
		edgeFrom = e.vertex1;
		edgeTo = e.vertex2;
	} else {
		edgeFrom = e.vertex2;
		edgeTo = e.vertex1;
	}
	os << "(" << edgeFrom << ", " << edgeTo << ")";
	return os;
}

File: python_parallel/kgm/edge.py
class Edge:
    """A single edge in an undirected graph.
    """

    def __init__(self, vertex1, vertex2):
        """Creates a new edge. Order of the vertices doesn't matter.
        :param vertex1: number of the first vertex (0 - (length of graph - 1))
        :param vertex2: number of the second vertex1 (0 - (length of graph - 1))
        """
        if vertex1 > vertex2:
            vertex1, vertex2 = vertex2, vertex1
        self.vertex1 = vertex1
        self.vertex2 = vertex2

    def isBacktrackMarker(self):
        """Minimum spanning tree algorithm specific method.
        :return: True if the edge is an instance of the backtrack market.
        """
        return self.vertex1 == -1 and self.vertex2 == -1

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return "({0}, {1})".format(self.vertex1, self.vertex2)

    def __eq__(self, other):
        sameOrder = self.vertex1 == other.vertex1 and self.vertex2 == other.vertex2
        crossedOrder = self.vertex1 == other.vertex2 and self.vertex2 == other.vertex1
        return sameOrder or crossedOrder

    def __lt__(self, other):
        return (10 * self.vertex1 + self.vertex2) < (10 * other.vertex1 + other.vertex2)

    def __ne__(self, other):
        return not self.__eq__(other)

    def __hash__(self):
        return hash(self.vertex1) * hash(self.vertex2)
